server/common/config/base_config.py

```diff
@@ -0,0 +1,104 @@
+import copy
+
+from flatten_dict import flatten
+from server.common.errors import ConfigurationError
+
+
+class BaseConfig(object):
+    """This class handles the mechanics of updating and checking attributes.
+    Derived classes are expected to store the actual attributes"""
+
+    def __init__(self, app_config, default_config, dictval_cases={}):
+        # reference back to the app_config
+        self.app_config = app_config
+        # the complete set of attribute and their default values (unflattened)
+        self.default_config = default_config
+        # attributes where the value may be a dict (and therefore are not flattened)
+        self.dictval_cases = dictval_cases
+        # used to make sure every attribute value is checked
+        self.attr_checked = {key_name: False for key_name in self.create_mapping(default_config).keys()}
+
+    def create_mapping(self, config):
+        """Create a mapping from attribute names to (location in the config tree, value)"""
```

Comment:
I actually have no idea what this is doing, still. What is "location in the config tree". Why is the config a tree and what do the nodes in the tree represent?

Reply:
Yes, the comment definitely can be greatly improved on.

The purpose of this function is to create a dictionary that looks like this:

The keys are the name of attributes, using the double underscore convention.
For example:  authentication__type.

The values are a tuple. 
The first item of the tuple is a tuple of path elements.   I'm calling this a "location in the config tree" in the comment.
The second item of the tuple is the value of the config parameter.
For example:   (('authentication', 'type'), 'session'))

So, basically the keys are the attribute names.  
The values says where in config hierarchy you'd find that config parameter, and it's current value.